burn-wgpu/src/template/conv/conv_transpose2d.wgsl

```diff
@@ -0,0 +1,115 @@
+    let input_stride_1 = info[1];
+    let input_stride_2 = info[2];
+    let input_stride_3 = info[3];
+    let output_stride_0 = info[4];
+    let output_stride_1 = info[5];
+    let output_stride_2 = info[6];
+    let output_stride_3 = info[7];
+    let weight_stride_0 = info[8];
+    let weight_stride_1 = info[9];
+    let weight_stride_2 = info[10];
+    let weight_stride_3 = info[11];
+
+    let input_shape_0 = info[12];
+    let input_shape_1 = info[13];
+    let input_shape_2 = info[14];
+    let input_shape_3 = info[15];
+    let output_shape_0 = info[16];
+    let output_shape_1 = info[17];
+    let output_shape_2 = info[18];
+    let output_shape_3 = info[19];
+    let weight_shape_0 = info[20];
+    let weight_shape_1 = info[21];
+    let weight_shape_2 = info[22];
+    let weight_shape_3 = info[23];
+
+    let stride_0 = info[24];
+    let stride_1 = info[25];
+    let padding_0 = info[26];
+    let padding_1 = info[27];
+    let dilation_0 = info[28];
+    let dilation_1 = info[29];
+    let groups = info[30];
+
+    let in_channels = weight_shape_0;
+    let kernel_size_0 = weight_shape_2;
+    let kernel_size_1 = weight_shape_3;
+
+    let b = id / output_stride_0 % output_shape_0;
+    let oc_out = id / output_stride_1 % output_shape_1;
+    let oh = id / output_stride_2 % output_shape_2;
+    let ow = id / output_stride_3 % output_shape_3;
+
+    let k = oc_out / weight_shape_1;
+    let g = k % groups;
+    let oc = oc_out - (weight_shape_1 * g);
+
+    var sum = bias[oc_out];
+
+    let ic_start = g * (in_channels / groups);
+    let ic_end = ic_start + in_channels / groups;
+
+    let ih_start = 0u;
+    let ih_end = input_shape_2;
+
+    let iw_start = 0u;
+    let iw_end = input_shape_3;
+
+    for (var ic = ic_start; ic < ic_end; ic++) {
+        for (var ih = ih_start; ih < ih_end; ih++) {
+            for (var iw = iw_start; iw < iw_end; iw++) {
+                for (var kh = 0u; kh < kernel_size_0; kh++) {
+                    for (var kw = 0u; kw < kernel_size_1; kw++) {
+                        let oh_tmp = ih * stride_0 + kh * dilation_0;
+                        let ow_tmp = iw * stride_1 + kw * dilation_1;
+
+                        if oh_tmp >= padding_0 && ow_tmp >= padding_1 {
+                            let oh_tmp_pad = oh_tmp - padding_0;
+                            let ow_tmp_pad = ow_tmp - padding_1;
+
+                            if oh_tmp_pad == oh && ow_tmp_pad == ow {
+                                let index_input = b * input_stride_0 + ic * input_stride_1 + ih * input_stride_2 + iw * input_stride_3;
+                                let index_weight = ic * weight_stride_0 + oc * weight_stride_1 + kh * weight_stride_2 + kw * weight_stride_3;
+
+                                sum += input[index_input] * weight[index_weight];
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    output[id] = sum;
```

Comment:
Should we mention this is not the most optimal alg and there are rooms for improvements?

Reply:
I'll add issues for performance!